File: app/src/main/java/com/hiddenramblings/tagmo/GlideTagModule.java
package com.hiddenramblings.tagmo;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import androidx.annotation.NonNull;

import com.bumptech.glide.GlideBuilder;
import com.bumptech.glide.annotation.GlideModule;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.cache.ExternalPreferredCacheDiskCacheFactory;
import com.bumptech.glide.module.AppGlideModule;
import com.bumptech.glide.request.RequestOptions;
import com.hiddenramblings.tagmo.browser.Preferences;
import com.hiddenramblings.tagmo.browser.SettingsFragment;

@GlideModule
public class GlideTagModule extends AppGlideModule {

    public static final String IMAGE_NETWORK_NEVER = "NEVER";
    public static final String IMAGE_NETWORK_WIFI = "WIFI_ONLY";
    public static final String IMAGE_NETWORK_ALWAYS = "ALWAYS";

    public RequestOptions onlyRetrieveFromCache(Context context, RequestOptions requestOptions) {
        String imageNetworkSetting = new Preferences(context.getApplicationContext()).image_network();
        if (IMAGE_NETWORK_NEVER.equals(imageNetworkSetting)) {
            return requestOptions.onlyRetrieveFromCache(true);
        } else if (IMAGE_NETWORK_WIFI.equals(imageNetworkSetting)) {
            ConnectivityManager cm = (ConnectivityManager) context
                    .getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
            return requestOptions.onlyRetrieveFromCache(null == activeNetwork
                    || activeNetwork.getType() != ConnectivityManager.TYPE_WIFI);
        } else {
            return requestOptions.onlyRetrieveFromCache(false);
        }
    }

    @Override
    public void applyOptions(@NonNull Context context, GlideBuilder builder) {
        int diskCacheSizeBytes = 1024 * 1024 * 128; // 128 MB
        // The current size of the API image repo is 117.1 MB
        builder.setDiskCache(new ExternalPreferredCacheDiskCacheFactory(
                context, diskCacheSizeBytes));
        builder.setLogLevel(Log.ERROR);
        RequestOptions requestOptions = onlyRetrieveFromCache(context,
                new RequestOptions().diskCacheStrategy(DiskCacheStrategy.RESOURCE));
        builder.setDefaultRequestOptions(requestOptions);
    }

    @Override
    public boolean isManifestParsingEnabled() {
        return false;
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/GlideTagModule.kt
package com.hiddenramblings.tagmo

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.util.Log
import com.bumptech.glide.GlideBuilder
import com.bumptech.glide.annotation.GlideModule
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.cache.ExternalPreferredCacheDiskCacheFactory
import com.bumptech.glide.module.AppGlideModule
import com.bumptech.glide.request.RequestOptions
import com.hiddenramblings.tagmo.eightbit.os.Version

@GlideModule
class GlideTagModule : AppGlideModule() {
    private fun isConnectionWiFi(context: Context): Boolean {
        var result = false
        (context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager).run {
            if (Version.isMarshmallow) {
                getNetworkCapabilities(activeNetwork)?.run {
                    result = when {
                        hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true
                        hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> false
                        hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> true
                        else -> false
                    }
                }
            } else {
                @Suppress("DEPRECATION")
                activeNetworkInfo?.run {
                    result = type == ConnectivityManager.TYPE_WIFI
                            || type == ConnectivityManager.TYPE_ETHERNET
                }
            }
        }
        return result
    }


    private fun onlyRetrieveFromCache(context: Context, requestOptions: RequestOptions): RequestOptions {
        return when (Preferences(context.applicationContext).imageNetwork()) {
            IMAGE_NETWORK_NEVER -> { requestOptions.onlyRetrieveFromCache(true) }
            IMAGE_NETWORK_WIFI -> { requestOptions.onlyRetrieveFromCache(!isConnectionWiFi(context)) }
            else -> { requestOptions.onlyRetrieveFromCache(false) }
        }
    }

    override fun applyOptions(context: Context, builder: GlideBuilder) {
        // Current size of AmiiboAPI images folder is 131 MB
        val diskCacheSizeBytes = 1024 * 1024 * 192 // 192 MB
        builder.setDiskCache(ExternalPreferredCacheDiskCacheFactory(
            context, diskCacheSizeBytes.toLong()
        ))
        builder.setLogLevel(Log.ERROR)
        builder.setDefaultRequestOptions(onlyRetrieveFromCache(
            context.applicationContext,
            RequestOptions().diskCacheStrategy(DiskCacheStrategy.RESOURCE)
        ))
    }

    override fun isManifestParsingEnabled(): Boolean {
        return false
    }

    companion object {
        const val IMAGE_NETWORK_NEVER = "NEVER"
        const val IMAGE_NETWORK_WIFI = "WIFI_ONLY"
        const val IMAGE_NETWORK_ALWAYS = "ALWAYS"
    }
}